### ISSUE TITLE
Stack overflow in combined_mutation_reader

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#2908]
*Installation details*
Scylla version (or git commit hash): master

```
(gdb) bt
#0  dht::tri_compare (t1=..., t2=...) at dht/i_partitioner.cc:118
#1  0x00000000010e8680 in dht::tri_compare (t2=..., t1=...) at dht/i_partitioner.cc:456
#2  dht::token_comparator::operator() (this=<optimized out>, t1=..., t2=...) at dht/i_partitioner.cc:455
#3  0x0000000000ae2d32 in wrapping_range<dht::token>::is_wrap_around<dht::token_comparator&> (cmp=..., this=<optimized out>)
    at ./range.hh:220
#4  wrapping_range<dht::token>::after<dht::token_comparator> (cmp=..., point=..., this=<optimized out>) at ./range.hh:140
#5  sstables::<lambda(range<dht::token>&, const dht::token&)>::operator() (__closure=<optimized out>, b=..., a=...)
    at sstables/compaction.cc:95
#6  __gnu_cxx::__ops::_Iter_comp_val<sstables::belongs_to_current_node(const dht::token&, const token_range_vector&)::<lambda(range<dht::token>&, const dht::token&)> >::operator()<__gnu_cxx::__normal_iterator<const nonwrapping_range<dht::token>*, std::vector<nonwrapping_range<dht::token> > >, const dht::token> (this=<optimized out>, __val=..., __it=...) at /usr/include/c++/7/bits/predefined_ops.h:177
#7  std::__lower_bound<__gnu_cxx::__normal_iterator<const nonwrapping_range<dht::token>*, std::vector<nonwrapping_range<dht::token> > >, dht::token, __gnu_cxx::__ops::_Iter_comp_val<sstables::belongs_to_current_node(const dht::token&, const token_range_vector&)::<lambda(range<dht::token>&, const dht::token&)> > > (__val=..., __last=..., __first=..., __comp=...) at /usr/include/c++/7/bits/stl_algobase.h:959
#8  std::lower_bound<__gnu_cxx::__normal_iterator<const nonwrapping_range<dht::token>*, std::vector<nonwrapping_range<dht::token> > >, dht::token, sstables::belongs_to_current_node(const dht::token&, const token_range_vector&)::<lambda(range<dht::token>&, const dht::token&)> > (
    __val=..., __last=..., __first=..., __comp=...) at /usr/include/c++/7/bits/stl_algo.h:2034
#9  sstables::belongs_to_current_node (sorted_owned_ranges=..., t=...) at sstables/compaction.cc:95
#10 sstables::cleanup_compaction::filter_func() const::{lambda(streamed_mutation const&)#1}::operator()(streamed_mutation const&) const (
    sm=..., __closure=<optimized out>) at sstables/compaction.cc:426
#11 std::_Function_handler<bool (streamed_mutation const&), sstables::cleanup_compaction::filter_func() const::{lambda(streamed_mutation const&)#1}>::_M_invoke(std::_Any_data const&, streamed_mutation const&) (__functor=..., __args#0=...)
    at /usr/include/c++/7/bits/std_function.h:302
#12 0x0000000000aeb1fd in wrapping_range<dht::token>::before<dht::token_comparator&> (cmp=..., point=..., this=0x6010051a1328)
    at ./range.hh:123
#13 nonwrapping_range<dht::token>::before<dht::token_comparator&> (cmp=..., point=..., this=0x6010051a1328) at ./range.hh:460
#14 nonwrapping_range<dht::token>::contains<dht::token_comparator> (cmp=..., point=..., this=0x6010051a1328) at ./range.hh:511
#15 sstables::sstable_set::incremental_selector::select (this=this@entry=0x6010051a1320, t=...) at sstables/compaction_strategy.cc:141
#16 0x0000000000896739 in incremental_reader_selector::create_new_readers (this=0x6010051a12c0, t=<optimized out>) at database.cc:441
#17 0x0000000000999f9e in combined_mutation_reader::maybe_add_readers (this=this@entry=0x60100529ef80, t=0x0) at mutation_reader.cc:65
#18 0x000000000099c9f5 in combined_mutation_reader::prepare_next (this=this@entry=0x60100529ef80) at mutation_reader.cc:85
#19 0x000000000099fe4e in combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#20 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
#21 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#22 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44
#23 seastar::futurize<seastar::future<optimized_optional<streamed_mutation> > >::apply<combined_mutation_reader::next()::<lambda()> > (
    args=..., func=...) at ./seastar/core/future.hh:1302
#24 seastar::future<>::then<combined_mutation_reader::next()::<lambda()> > (func=..., this=0x7ffff0c00780) at ./seastar/core/future.hh:890
#25 combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#26 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
---Type <return> to continue, or q <return> to quit---
#27 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#28 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44
#29 seastar::futurize<seastar::future<optimized_optional<streamed_mutation> > >::apply<combined_mutation_reader::next()::<lambda()> > (
    args=..., func=...) at ./seastar/core/future.hh:1302
#30 seastar::future<>::then<combined_mutation_reader::next()::<lambda()> > (func=..., this=0x7ffff0c00870) at ./seastar/core/future.hh:890
#31 combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#32 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
#33 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#34 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44
#35 seastar::futurize<seastar::future<optimized_optional<streamed_mutation> > >::apply<combined_mutation_reader::next()::<lambda()> > (
    args=..., func=...) at ./seastar/core/future.hh:1302
#36 seastar::future<>::then<combined_mutation_reader::next()::<lambda()> > (func=..., this=0x7ffff0c00960) at ./seastar/core/future.hh:890
#37 combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#38 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
#39 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#40 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44
#41 seastar::futurize<seastar::future<optimized_optional<streamed_mutation> > >::apply<combined_mutation_reader::next()::<lambda()> > (
    args=..., func=...) at ./seastar/core/future.hh:1302
#42 seastar::future<>::then<combined_mutation_reader::next()::<lambda()> > (func=..., this=0x7ffff0c00a50) at ./seastar/core/future.hh:890
#43 combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#44 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
#45 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#46 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44
#47 seastar::futurize<seastar::future<optimized_optional<streamed_mutation> > >::apply<combined_mutation_reader::next()::<lambda()> > (
    args=..., func=...) at ./seastar/core/future.hh:1302
#48 seastar::future<>::then<combined_mutation_reader::next()::<lambda()> > (func=..., this=0x7ffff0c00b40) at ./seastar/core/future.hh:890
#49 combined_mutation_reader::next (this=this@entry=0x60100529ef80) at mutation_reader.cc:103
#50 0x00000000009a047a in combined_mutation_reader::<lambda()>::operator() (__closure=<optimized out>) at mutation_reader.cc:103
#51 seastar::apply_helper<combined_mutation_reader::next()::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=..., func=...) at ./seastar/core/apply.hh:36
#52 seastar::apply<combined_mutation_reader::next()::<lambda()> > (args=..., func=...) at ./seastar/core/apply.hh:44

```

**Comment by tgrabiec** [comment]
@raphaelsc Is this the one you're working on?

**Comment by raphaelsc** [comment]
On Tue, Oct 17, 2017 at 6:06 PM, Tomasz Grabiec <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc> Is this the one you're working
> on?
>

yes

**Comment by slivne** [comment]
@raphaelsc ping

**Comment by slivne** [comment]
@raphaelsc - I can see that you are working on a patch and tomek had comments 

Is this a regression from 2.0/1.7 ?

If not we will pus this to 2.2

**Comment by raphaelsc** [comment]
On Sun, Dec 31, 2017 at 9:13 AM, Shlomi Livne <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc> - I can see that you are
> working on a patch and tomek had comments
>
> Is this a regression from 2.0/1.7 ?
>
> If not we will pus this to 2.2
>


@slivne: same as @2863

branch-2.1 contains the regression, so a fix is needed for 2.1. New version
of patchset to fix it is on the mailing list.
